tests/integration-service/namespace-backed-environments.go

```diff
@@ -0,0 +1,311 @@
+			timeout = time.Minute * 10
+			Eventually(func() error {
+				pipelineRun, err = f.AsKubeAdmin.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
+				if err != nil {
+					GinkgoWriter.Printf("PipelineRun has not been created yet for Component %s/%s\n", testNamespace, componentName)
+					return err
+				}
+				if !pipelineRun.HasStarted() {
+					return fmt.Errorf("pipelinerun %s/%s hasn't started yet", pipelineRun.GetNamespace(), pipelineRun.GetName())
+				}
+				return nil
+			}, timeout, constants.PipelineRunPollingInterval).Should(Succeed(), fmt.Sprintf("timed out when waiting for the PipelineRun to start for the component %s/%s", testNamespace, componentName))
+			Expect(f.AsKubeDeveloper.HasController.WaitForComponentPipelineToBeFinished(originalComponent, "", 2, f.AsKubeAdmin.TektonController)).To(Succeed())
+			Expect(pipelineRun.Annotations["appstudio.openshift.io/snapshot"]).To(Equal(""))
+		}
+
+		assertBuildPipelineRunSigned := func() {
+			Eventually(func() error {
+				pipelineRun, err := f.AsKubeAdmin.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
+				if err != nil {
+					GinkgoWriter.Printf("PipelineRun for Component %s/%s can't be gotten successfully\n", testNamespace, componentName)
+					return err
+				}
+				if pipelineRun.Annotations["chains.tekton.dev/signed"] != "true" {
+					return fmt.Errorf("pipelinerun %s/%s hasn't been signed yet", pipelineRun.GetNamespace(), pipelineRun.GetName())
+				}
+				return nil
+			}, timeout, constants.PipelineRunPollingInterval).Should(Succeed(), fmt.Sprintf("timed out when waiting for the PipelineRun to be signed for component %s/%s", testNamespace, componentName))
+		}
+
+		assertSnapshotCreated := func() {
+			Eventually(func() error {
+				snapshot, err = f.AsKubeAdmin.IntegrationController.GetSnapshot("", "", componentName, testNamespace)
+				if err != nil {
+					GinkgoWriter.Printf("cannot get the Snapshot: %v\n", err)
+					return err
+				}
+				return nil
+			}, timeout, interval).Should(Succeed(), fmt.Sprintf("timed out when trying to check if the Snapshot for component %s/%s exists", testNamespace, componentName))
+			Eventually(func() error {
+				pipelineRun, err := f.AsKubeAdmin.IntegrationController.GetBuildPipelineRun(componentName, applicationName, testNamespace, false, "")
+				if err != nil {
+					GinkgoWriter.Printf("PipelineRun for Component %s/%s can't be gotten successfully\n", testNamespace, componentName)
+					return err
+				}
+				if pipelineRun.Annotations["appstudio.openshift.io/snapshot"] != snapshot.GetName() {
+					return fmt.Errorf("build pipelinerun %s/%s hasn't been annotated with snapshot name", pipelineRun.GetNamespace(), pipelineRun.GetName())
+				}
+				return nil
+			}, timeout, interval).Should(Succeed(), fmt.Sprintf("timed out when waiting for the buildPipelineRun annotated with Snapshot name %s", snapshot.GetName()))
+		}
```

Comment:
there's a lot of code duplication in these types of functions in this test file and [integration.go](https://github.com/redhat-appstudio/e2e-tests/blob/f606434eb886404ae5e5da7d4dc130cc0dde41e7/tests/integration-service/integration.go#L129-L149). I'm wondering if they could be written in a way that they are shared at least for both integration-service tests

Reply:
Agreed. Out of the functions: `createApp`, `createComponent`, `cleanup`, `assertBuildPipelineRunFinished`, `assertBuildPipelineRunSigned`, `assertSnapshotCreated`, `assertIntegrationPipelineRunFinished`, I've removed the last 4 from our test suite and replaced them with common/shared functions.

I didn't remove `createApp`, `createComponent`, and `cleanup` because I couldn't find a good place to put them in and it seemed like there would be more code duplication if we removed these methods instead. I'm open to suggestions!